### PR TITLE
ID Library

### PR DESCRIPTION
**Pull request by SKOCHERI**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer eng-dmp@magnite.com
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Comment by jdwieland8282** [comment]
docs pr: https://github.com/prebid/prebid.github.io/pull/2429

**Comment by smenzer** [comment]
@SKOCHERI or @jdwieland8282 can you provide a bit of context as to what this is for, please?

**Comment by jdwieland8282** [comment]
@smenzer context is in the docs pr, but specifically to answer your question:

"The ID Library module gathers and generates a map of identities present on the page.  The primary usecase for this adapter is for Publishers who have included multiple UserId subadapters in their prebid.js implementation, and want to store the resulting user ids server side for modeling or graphing purposes. "

**Comment by smenzer** [comment]
@jdwieland8282 ah thanks...I opened the docs PR looking for some context but didn't actually go read the .md file 🤦 

from the docs and code, it looks like this is all tied to a hashed email, is that correct? what if there isn't an email available, can the publisher provide their own UID or something?

**Comment by jdwieland8282** [comment]
its all tied to a persistent identifier, hashed email is one option but not the only option. If there id library module doesn't find an email nothing happens.  The publisher can declare an element id for where they have store the users userName on the page:  

ex. "2. a publisher configured element, for example when configuration `target='username'`, the value is extracted from an element having id `'username'`"

**Review by robertrmartinez — CHANGES_REQUESTED** [review]
Still need to do some more review but had a few concerns / comments

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js:55]: Unless the publisher has configured the element from which the email needs to be fetched, we will have to scan the entire page to get email.

**Comment by jdwieland8282** [comment]
adding some context behind the motivation for creating this module

_Our point of view with regard to identity is that a 1st party relationship between publishers and viewers of their content exists.  As such, we believe that publishers have a right to anchor all user ids generated by the Prebid userid module, to a persistent (hashed) identifier such as an email address. We believe that it is in the publisher, and by extension the open web’s interest, to create a deterministic id graph composed of the linkage between persistent ids and 1st party ids generated by the user id sub-adapters. As browsers limit our ability to generate cross domain identifiers, we believe a publisher controlled id graph is the right response.  We (Magnite) may decide to help publishers build this graph in the future if asked._

File: modules/idLibrary.js
Comment (smenzer) [reply modules/idLibrary.js:55]: why can't the publisher just pass the hashed email directly to the library when they initiate it? i agree that scanning an entire page is not ideal

Comment (robertrmartinez) [reply modules/idLibrary.js:51]: I recommend we make this scanning of entire document a configurable option via the `setConfig` call to this module.

We will leave it defaulting to `true` but I think it is in our best interest to document the behavior of the module thoroughly, and make sure the publisher has the control necessary.

@jdwieland8282 Can we make sure the associated docs PR contains the steps outlined and the fallbacks?

Something that outlines the flow:

How the target is found:

1. Uses publisher defined target element
2. Searches for HTML input (text/email) element
3. Searches entire document for email using regex.

And outline that the latter (3) can be configured off by something like:

```js
pbjs.setConfig({
   idLibrary: {
      fallbackSearch: false // defaults to true (I am bad at naming vars)
   }
});
```

@SKOCHERI

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:112]: So debounce is just returning a function to be called later?

However, anywhere debounce is called, the return is not used.

So this is effectively not doing anything.

I think the real fix here is to add the method bindings as requested for the other methods (first param anytime debounce is called).

And to make this function a regular function that just executes, not one which returns a function.

Comment (robertrmartinez) [reply modules/idLibrary.js:112]: Also, did you mean to use the keyword `arguments` instead of `args` ?

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:132]: @SKOCHERI 

What is the intent here? Please elaborate on the expectation of things to happen when this line is hit.

`debounce` takes in a `func` param that looks like the intent is for it to be called later.

But you are executing the `targetAction` here with those params, which returns undefined. So really `func` is always undefined when `debounce` is run.

Is the intent to not call `targetAction` here but just pass its reference to debounce?

If so, recommend you use `bind`.

```js
debounce(targetAction.bind(null, conf, mutations, observer), conf.debounce, false);
```

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:167]: Same as `debounce` change needed here.

**Review by robertrmartinez — CHANGES_REQUESTED** [review]
There's some changes needed.

File: modules/idLibrary.js
Comment (jdwieland8282) [reply modules/idLibrary.js:51]: @robertrmartinez the existing Docs PR describes the following:

_A persistent identifier can be extracted in the following ways:

1. From a generic `<div>` element
2. From a publisher configured HTML element id
3. From an `<input>` element of type text/email_

Are you suggesting **your** 1-3 above replace the existing 1-3?

Comment (robertrmartinez) [reply modules/idLibrary.js:51]: Well just something that outlines exactly the flow of the module.

Could be another flow chart or something.

**Comment by smenzer** [comment]
@jdwieland8282 is there a use case where the publisher doesn't want this library to scan the page for a hashed email, but has their own identifier already (maybe in a 1P cookie, session, etc) but they still want to collect all of the userId module IDs and post them to themselves? in other words...is it possible to turn off the first 4 steps in "Application Flow" (https://github.com/prebid/prebid.github.io/pull/2429/files) and just capture and post the user IDs someplace?

**Comment by jdwieland8282** [comment]
Hi @smenzer , Not exactly, the intention here is to make the output deterministic in nature, and not require a bunch of probabilistic ML.  If a user wanted to do what you suggested, they could deploy the pubprovided userid module, but they would be limited to page loads where a persistent value was observed.

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: This has been fixed

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: @robertrmartinez The debounce issue is fixed and full body scan config is provided.

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: The debounce issue is fixed

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:125]: I think it would make more sense to instead just wrap this logic in a if else

if (callNow)

  call the func

else

   setTimeout to call the func.

Unnecessarily executing a setTimeout when !immediate seems like a waste.

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:105]: same, strict check since getEmail explicitly will return null.

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:88]: Could use a `forEach` instead of a for loop since you do not need to exit early or anything.

Just a personal preference, can leave it.

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:117]: usually we use let / or const now instead of actual var declaration.

No big deal though.

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:70]: If we get an email do we need to keep going through all other mutations?

Should there only ever be one email on page?

Should we break here?

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:56]: !==

**Review by robertrmartinez — CHANGES_REQUESTED** [review]
Several more questions / comments

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js:193]: refreshUserIds is called prior to posting data which happens only when a email is found.

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Fixed

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Fixed

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: userIds is an object, so updated the check to Object.keys(userIds).length === 0

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Removed

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: fixed

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js:89]: using forEach LGTM analysis: JavaScript test failed so used for loop

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Updated

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js:70]: Once email is found, data is posted and returned now

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Fixed

**Comment by SKOCHERI** [comment]
@robertrmartinez @smenzer All the review comments has been addressed. Can you take a look and approve if no more comments.

**Comment by smenzer** [comment]
I have no other comments, thanks.

**Review by smenzer — APPROVED** [review]
LGTM, pending review by @robertrmartinez

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:145]: I think the space is already included in the logInfo function you created, are these needed?

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:161]: unnecessary space?

File: modules/idLibrary.js
Comment (robertrmartinez) [inline modules/idLibrary.js:191]: spaces still

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Fixed

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Fixed

File: modules/idLibrary.js
Comment (SKOCHERI) [reply modules/idLibrary.js]: Fixed

**Comment by robertrmartinez** [comment]
Code looks good now,

I am going to do some on page testing today and if it goes well we should be good to merge for tomorrows release!

**Review by robertrmartinez — APPROVED** [review]

**Pull request merged**
